### PR TITLE
Bug/target label unmarshal

### PR DESCRIPTION
**Pull request by Nomon**
Fix an issue that prevented prometheus from starting up when using target_label interpolation in replace action due to it failing LabelName regex validation.

Changes:
- Change the TargetLabel to string to avoid it being Unmarshaled and validated as a LabelName
- Validate the TargetLabel with a different regex if its a replace action, with old if not.

Refs:
https://github.com/prometheus/prometheus/pull/2078#issuecomment-254539237
https://github.com/prometheus/prometheus/pull/2090

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -988,7 +988,8 @@ type RelabelConfig struct {
 	// Modulus to take of the hash of concatenated values from the source labels.
 	Modulus uint64 `yaml:"modulus,omitempty"`
 	// The label to which the resulting string is written in a replacement.
-	TargetLabel model.LabelName `yaml:"target_label,omitempty"`
+	// regex interpolation is allowed for the replace action.
```
Comment (brancz) [inline config/config.go]: Capitalize beginning of a comment.

File: config/testdata/conf.good.yml [hunk 2]
```diff
@@ -104,6 +104,13 @@ scrape_configs:
   - server: 'localhost:1234'
     services: ['nginx', 'cache', 'mysql']
 
+  relabel_configs:
+  - source_labels: [__meta_sd_consul_tags]
+    separator: ','
```
Comment (brancz) [inline config/testdata/conf.good.yml]: Could you even this out with spaces as well?

File: config/config.go [hunk 3]
```diff
@@ -1014,6 +1015,12 @@ func (c *RelabelConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	if (c.Action == RelabelReplace || c.Action == RelabelHashMod) && c.TargetLabel == "" {
 		return fmt.Errorf("relabel configuration for %s action requires 'target_label' value", c.Action)
 	}
+	if c.Action == RelabelReplace && !relabelTarget.Match([]byte(c.TargetLabel)) {
+		return fmt.Errorf("%q is invalid 'target_label' for %s action", c.TargetLabel, c.Action)
+	}
+	if c.Action == RelabelHashMod && !model.LabelNameRE.Match([]byte(c.TargetLabel)) {
```
Comment (brancz) [inline config/config.go]: Hmm.. I think I understand a bit better now, I'm unsure whether it is possible for us to validate at this point when we allow interpolation, as we don't know if for example the interpolation causes the label to start with `[a-zA-Z]`.

**Comment by brancz** [comment]
Could @beorn7 and/or @brian-brazil review this? I feel like I'm missing the context a bit, and as you reviewed the previous PR (and know the surrounding code better) you might have better insight.

**Comment by beorn7** [comment]
I'll have a look ASAP. Hopefully later today.

File: config/config.go (hunk 3 above)
Comment (Nomon) [inline config/config.go]: The relabelTarget regex makes sure that the label starts with a-zA-Z_ and containing a-zA-Z_0-9 after that while allowing regexp interpolations at any position. If its RelabelHashMod the old LabelNameRE is used as no interpolation is supported for that action.

Comment (Nomon) [inline config/config.go]: The isValid method is used in the relabel.go, this validates that the resulting target_label is a valid label name ([a-zA-Z_][a-zA-Z0-9_]*)

File: config/config.go [hunk 4]
```diff
@@ -1014,6 +1015,12 @@ func (c *RelabelConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	if (c.Action == RelabelReplace || c.Action == RelabelHashMod) && c.TargetLabel == "" {
 		return fmt.Errorf("relabel configuration for %s action requires 'target_label' value", c.Action)
 	}
+	if c.Action == RelabelReplace && !relabelTarget.Match([]byte(c.TargetLabel)) {
```
Comment (beorn7) [inline config/config.go]: The regexp type has a method `MatchString`, which you can use here.

File: config/config.go (hunk 3 above)
Comment (beorn7) [inline config/config.go]: Use `MatchString` as above.

Comment (beorn7) [inline config/config.go]: But at this point, where you have no interpolation, you can indeed yous the `IsValid` method @brancz pointed to. It is doing the same as matching against `LabelNameRE`, just faster (which isn't critical here, but it's still more consistent).

Comment (Nomon) [inline config/config.go]: I used the regexp here so it does the same validation as it did before (in unmarshalyaml). I can swap it to the isValid

**Comment by beorn7** [comment]
Very nice. Thanks for the fix.

**Pull request merged**